### DISCUSSION TITLE
Intersperse views into code

### DISCUSSION THREAD

**Discussion by Ivorforce** [microsoft/vscode-discussions#592] [Extension Development QnA] [answered]
Hiya, I'm looking for a way to intersperse some custom views / html into the code editor.

Code Lens (e.g. GitLens) do it almost perfectly:
![SCR-20230504-33x](https://user-images.githubusercontent.com/2308167/236077435-4dc88f8e-611d-41aa-85cb-117e8d2e78f4.png)

Notice how the line numbers move to give room to the lens actions.

---

However, its capabilities don't quite extend far enough. I need to be able to adjust the height of the interspersed view (for multiline additions), as well as insert more advanced HTML (rather than just clickable plaintext).
Here is an example:

![example](https://user-images.githubusercontent.com/2308167/236079201-9bb64d1d-9b40-4089-9dd2-92ad67291d98.png)

Here, I would give existing text a red background, to indicate it would be removed in a git diff. Meanwhile, I'm interspersing different code (just rendered as text, syntax hightlighting optional) to indicate it will be added in the same diff.

The only solution I can think of is using git lens as a framework to insert views where I need them, then access the HTML directly to modify it until it fits my needs. That feels a bit hacky though.

Does anybody have a solution? I'd be more than happy if somebody simply knew an extension that does something similar, so that I can browse their source code.

Cheers,

Ivorius

**Comment by alefragnani** [comment, marked as answer]
Hi @Ivorforce ,

This is provided by the CodeLens API (https://github.com/microsoft/vscode-extension-samples/tree/main/codelens-sample / https://code.visualstudio.com/api/language-extensions/programmatic-language-features#codelens-show-actionable-context-information-within-source-code), and it only supports displaying text, as you see in GitLens.

Based on your description, I suppose what you need is being asked in https://github.com/microsoft/vscode/issues/3220 (custom peek widgets).

Hope this helps

**Reply by Ivorforce** [reply]
The peek widgets are pretty close to what I'd need, thanks for the link!
Edit: The feature got me to [Editor Insets](https://github.com/microsoft/vscode/issues/85682), which is a feature that's available in VSCode Insiders. Nice!

**Comment by gjsjohnmurray** [comment]
Also ZoneWidget API request (#140752)
